### ISSUE TITLE
no-use-before-define should have an option for ES6 classes

### ISSUE THREAD

**Issue by lucsky** [eslint/eslint#3944]
The "no-use-before-define" rule has a "nofunc" option which allows to specify that using a function before it has been declared is ok. It should be possible (with a "noclass" option for example) to also do that with ES6 classes where it would allow a class name to be used before the class is declared later in the file.

Ex, this triggers the "Bar was used before it was defined" error:

```
class Foo {
    constructor() {
        this.bar = new Bar();
    }
}

class Bar {}
```

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by gyandeeps** [comment]
Duplicate https://github.com/eslint/eslint/issues/3941

**Comment by michaelficarra** [comment]
I guess this new option should apply to anything that hoists with TDZ. Sounds good, though definitely dangerous.

**Comment by nzakas** [comment]
Won't that code cause a runtime error?

**Comment by michaelficarra** [comment]
No, TDZ errors occur when the reference is evaluated before the declaration. So there's no way that this program would have a problem, regardless of what follows it.

**Comment by nzakas** [comment]
Oh right, because the reference doesn't occur until `new` is called. Got it.

**Comment by mysticatea** [comment]
I think "noclass" is confusable.
The following pattern is a runtime error different from functions.

```js
let a = new A();
class A { }
```

So I favor of "allow-outer-scope" or something like.

**Comment by nzakas** [comment]
Good point, though I don't think `allow-outer-scope` is clear enough. Is it even possible to accurately flag one but not the other in the error?

**Comment by mysticatea** [comment]
I'm sorry.

Currently, `no-use-before-define` has an string option `"nofunc"`.
We will add new option, so I think a use of an object is fine.

My proposal:

```js
{
    "no-use-before-define": [2, {"functions": true, "outerClasses": true}]
}
```

- `"functions"` (`boolean`) - If this option is `false`, `FunctionDeclaration`s are allowed to use before declarations. Default is `true`.
- `"outerClasses"` (`boolean`) - If this option is `false`, `ClassDeclaration`s are allowed to use before declarations only if the declaration is in the scope of an outer function (or global). Default is `true`.

```js
module.exports.schema = [
    {
        "oneOf": [
            {
                "enum": ["nofunc"]
            },
            {
                "type": "object",
                "properties": {
                    "functions": {"type": "boolean"},
                    "outerClasses": {"type": "boolean"}
                },
                "additionalProperties": false
            }
        ]
    }
];
```

**Comment by nzakas** [comment]
Why outer classes only?

**Comment by mysticatea** [comment]
@nzakas Class declarations are not hoisted, so the following code raises a runtime error.

```js
let a = new A();
class A { }
```

The following code is OK.

```js
module.exports = function foo() {
    let a = new A();
};

class A { }
```

**Comment by nzakas** [comment]
Yes, that part I understand, I just don't understand the term "outer classes". What are you calling an "outer class"?

**Comment by mysticatea** [comment]
I imaged "class declarations in upper function scopes or global scopes", and I saw a Ja-En dictionary.
I want to get advice.

**Comment by nzakas** [comment]
I think we can just use "classes" and have it work the way you describe. We just need to explain in the documentation.

**Comment by mysticatea** [comment]
OK.
I will work on this.
